File: leetcode-go/Q051-Q100/Q086_Partition_List_B/Q086_Partition_List_B.go
package Q086_Partition_List_B

type ListNode struct {
	Val  int
	Next *ListNode
}

func partition(head *ListNode, x int) *ListNode {
	dummy1, dummy2 := &ListNode{}, &ListNode{}
	cur1, cur2 := dummy1, dummy2

	for head != nil {
		if head.Val < x {
			cur1.Next = head
			cur1 = cur1.Next
		} else {
			cur2.Next = head
			cur2 = cur2.Next
		}
		head = head.Next
	}

	cur2.Next = nil
	cur1.Next = dummy2.Next
	return dummy1.Next
}

File: src/facebook/Q086_Partition_List_B.java
package facebook;

public class Q086_Partition_List_B {

    public ListNode partition(ListNode head, int x) {
        ListNode dummy1 = new ListNode(0);
        ListNode dummy2 = new ListNode(0);
        ListNode p1 = dummy1, p2 = dummy2;
        while (head != null) {
            ListNode next = head.next;
            head.next = null;
            if (head.val < x) {
                p1.next = head;
                p1 = p1.next;
            } else {
                p2.next = head;
                p2 = p2.next;
            }
            head = next;
        }
        p1.next = dummy2.next;
        return dummy1.next;
    }

    class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

}
